File: python/39_CombinationSum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        total_set = []
        def helper(cands, target, curr=[]):
            if target < 0 or len(cands) <= 0:
                return total_set
            elif target == 0:
                total_set.append(curr)
                return total_set
            else:
                helper(cands, target-cands[0], curr + [cands[0]])
                helper(cands[1:], target, curr)
            return total_set
        return helper(candidates, target, [])   

class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        if candidates == []:
            return []
        d = {}
		
        for candidate in candidates:
            for i in range(1, target+1):
                if i == candidate:
                    if i not in d:
                        d[i] = [[candidate]]
                    else:
                        d[i] += [[candidate]]
						
				# see if candidate-i  exists in dicationary already
                elif i-candidate > 0 and i-candidate in d:
                    for num_set in d[i-candidate]:
						# add candidate to the saved sets in d[i-candidate]
                        x = num_set + [candidate]
                        if i not in d:
                            d[i] = [x]
                        else:
                            d[i].append(x)
        if target not in d:
            return []
        else:
            return d[target]     

File: cpp/39_CombinationSum.cpp
class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        unordered_map<int, vector<vector<int>>> d = {};
        
        for (int cand : candidates) {
            for (int i = 0; i <= target; ++i) {
                if (i == cand) {
                    if (d.find(i) == d.end()) {
                        d[i] = {{i}};
                    } else {
                        d[i].push_back({cand});
                    }
                } else if (i-cand > 0 && d.find(i-cand) != d.end()) {
                    for (vector<int> num_set : d[i-cand]) {
                        num_set.push_back(cand);
                        if (d.find(i) == d.end()) {
                            d[i] = {num_set};
                        } else {
                            d[i].push_back({num_set});
                        }
                    }
                }
            }
        }
        return d.find(target) != d.end() ? d[target] : vector<vector<int>>{};
    }
};
